src/bdiff.c

```diff
@@ -78,32 +124,61 @@ hunk * const bdiff_narrow(
-            &precise_hunks_head, &precise_hunks_tail, rough_hunks->a.start,
-            rough_hunks->a.end, rough_hunks->b.start, rough_hunks->b.end);
-        ds(a, precise_hunks_tail->a.start);
-        ds(b, precise_hunks_tail->b.start);
-        char buf_a[buf_size], buf_b[buf_size];
+        if (end_shove_a) {
+            end_shove_a = slidey_aligner(
+                sample_size, ds, df, buf_a, buf_b, a, b,
+                rough_hunks->a.start, precise_hunks_tail->b.end,
+                min3(
+                    precise_hunks_tail->b.end - precise_hunks_tail->b.start,
+                    rough_hunks->a.end - rough_hunks->a.start,
+                    max_chunk_size));
+            precise_hunks_tail->b.end -= end_shove_a;
+        }
+        if (end_shove_b) {
+            end_shove_b = slidey_aligner(
+                sample_size, ds, df, buf_b, buf_a, b, a,
+                rough_hunks->b.start, precise_hunks_tail->a.end,
+                min3(
+                    precise_hunks_tail->a.end - precise_hunks_tail->a.start,
+                    rough_hunks->b.end - rough_hunks->b.start,
+                    max_chunk_size));
+            precise_hunks_tail->a.end -= end_shove_b;
+        }
+        ds(a, rough_hunks->a.start + end_shove_a);
+        ds(b, rough_hunks->b.start + end_shove_b);
         unsigned const start_delta = find_start_delta(
             df, sample_size, buf_a, buf_b, a, b);
-        precise_hunks_tail->a.start += start_delta;
-        precise_hunks_tail->b.start += start_delta;
-        unsigned end_shove = 0;
-        if (precise_hunks_tail->a.start > precise_hunks_tail->a.end) {
-            end_shove =
-                precise_hunks_tail->a.start - precise_hunks_tail->a.end;
+        end_shove_a += start_delta;
+        end_shove_b += start_delta;
+        if (
+                ((rough_hunks->b.start + end_shove_b) ==
+                    rough_hunks->b.end) &&
+                ((rough_hunks->a.start + end_shove_a) ==
+                    rough_hunks->a.end)) {
+            // Hunk that contains nothing in either
+            end_shove_a = end_shove_b = 0;
+            continue;
         }
-        if (precise_hunks_tail->b.start > precise_hunks_tail->b.end) {
-            end_shove =
-                precise_hunks_tail->b.start - precise_hunks_tail->b.end;
-        }
-        precise_hunks_tail->a.end += end_shove;
-        precise_hunks_tail->b.end += end_shove;
-        unsigned end_delta = min(
+        append_hunk(
+            &precise_hunks_head, &precise_hunks_tail,
+            rough_hunks->a.start + end_shove_a,
+            rough_hunks->a.end,
+            rough_hunks->b.start + end_shove_b,
+            rough_hunks->b.end);
+        end_shove_a = clamped_subtract(
+            precise_hunks_tail->a.start, precise_hunks_tail->a.end);
+        end_shove_b = clamped_subtract(
+            precise_hunks_tail->b.start, precise_hunks_tail->b.end);
+        assert(!(end_shove_a && end_shove_b));
```

Comment:
If we can't have non-zero values for both, is the conditional on line 140 really an `else if`?